model/state_model_dynamic.jl

```diff
@@ -0,0 +1,655 @@
+    
+        isempty(idx) && (idx = Dict(k => findindex(df) for (k,df) in d2))
+        (set, idx) = _model_set!(d2, set, idx)
+
+        d1 = Dict(k => filter_with(df, (yr = year,); extrapolate = false, drop = true) for (k,df) in d)
+        d1 = Dict(k => convert_type(Dict, fill_zero(set, df)) for (k,df) in d1)
+        return (d1, set, idx)
+end
+
+function _model_input(year::UnitRange{Int64}, d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict = Dict())
+        return _model_input(ensurearray(year), d, set, idx)
+end
+
+"""
+    _model_set!(d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict)
+This function returns subsets intended to limit the size of the model by including only
+non-zero values when mapping zero-profit and market-clearing conditions.
+
+# Arguments
+- `d::Dict{Symbol,DataFrame}` of DataFrames containing the model data.
+- `set::Dict` of Arrays describing region, sector, final demand, etc.
+- `idx::Dict` of parameter indices
+
+# Returns
+- `d::Dict{Symbol,Dict}` of model data with all zeros filled
+- `set::Dict` of model indices, with added arrays of tuples indicating which (region,sector)
+    or (region,good) combinations are relevant to the zero-profit and market-clearing
+    conditions.
+- `idx::Dict` of indices, updated to include those used to define the newly-added sets.
+"""
+function _model_set!(d::Dict{Symbol,DataFrame}, set::Dict, idx::Dict)
+    (set[:A], idx[:A]) = nonzero_subset(d[:a0] + d[:rx0])
+    (set[:Y], idx[:Y]) = nonzero_subset(combine_over(d[:ys0], :g))
+    (set[:X], idx[:X]) = nonzero_subset(d[:s0])
+    (set[:PA], idx[:PA]) = nonzero_subset(d[:a0])
+    (set[:PD], idx[:PD]) = nonzero_subset(d[:xd0])
+    (set[:PK], idx[:PK]) = nonzero_subset(d[:kd0])
+    (set[:PY], idx[:PY]) = nonzero_subset(d[:s0])
+    return (set, idx)
+end
+
+
+
+############
+# LOAD DATA
+############
+
+#SLiDE data needs to be built or point to pre-existing build directory
+#can pass a name (d, set) = build_data("name_of_build_directory")
+!(@isdefined(d_in) && @isdefined(set_in)) && ((d_in, set_in) = build_data("state_model"))
+d = copy(d_in)
+set = copy(set_in)
+
+#Specify benchmark year and end year for dynamic model time horizon
+bmkyr = 2016
+endyr = 2018
+
+#Define range of years in time horizon
+years = bmkyr:endyr
+
+#Load slide data and time horizon to produce model data and appropriate time-indexed subsets
+(sld, set, idx) = _model_input(years, d, set)
+
+#last year is the maximum of all years
+years_last = maximum(years)
+
+#Create boolean indicators of first and last year of time horizon
+bool_firstyear = Dict()
+bool_lastyear = Dict()
+for t in years
+        if t!=years_last
+                push!(bool_lastyear,t=>0)
+        else
+                push!(bool_lastyear,t=>1)
+        end
+
+        if t!=bmkyr
+                push!(bool_firstyear,t=>0)
+        else
+                push!(bool_firstyear,t=>1)
+        end
+end
```

Comment:
Adding checkboxes throughout my comments as well so that I can come back through them and mark complete when finished.

#### Run Note
Interesting, I don't get that warning when I run the state_model*.jl files. I am on julia 1.4.2 though, so maybe it is a versioning thing. That said, I do get errors regarding your two examples. But when I throw them into a function, I don't get any errors or warnings.

- [ ] Let's discuss


#### Concision
Yes we do use bool_firstyear. bmkyr will more than likely, always be the first year. Though I understand the confusion with the naming of bmkyr. For example, in the fully intertemporal model, all years are the benchmark. Let me think about this.

- [ ] Consider clarifying naming convention for bmkyr
- [ ] Define independently of bmkyr
- [ ] Define bools in one line

#### Style
Totally agree here - will implement this change. A common modeling notation is also to use something like "tf" and "tl" or "tfirst" and "tlast". "yrf" and "yrl" could be good as well, but for the bool, I like appending "is". I definitely don't like bool_firstyear. 

- [ ] Settle on naming convention and update code